pkgs/racket-doc/scribblings/reference/equality.scrbl

```diff
@@ -0,0 +1,306 @@
+ @racket[eqv?] differs from @racket[eq?]. Two numbers are @racket[eqv?] when
+ they have the same exactness, precision, and are both equal and non-zero, both
+ @racketvalfont{+0.0}, both @racketvalfont{+0.0f0}, both @racketvalfont{-0.0},
+ both @racketvalfont{-0.0f0}, both @racketvalfont{+nan.0}, or both
+ @racketvalfont{+nan.f}---considering real and imaginary components separately
+ in the case of @tech{complex numbers}. Two characters are @racket[eqv?] when
+ their @racket[char->integer] results are equal.
+
+ Generally, @racket[eqv?] is identical to @racket[equal?] except that the former
+ cannot recursively compare the contents of compound data types (such as lists
+ and structs) and cannot be customized by user-defined data types. The use of
+ @racket[eqv?] is lightly discouraged in favor of @racket[equal?].
+
+ @(examples
+   (eqv? 'yes 'yes)
+   (eqv? 'yes 'no)
+   (eqv? (* 6 7) 42)
+   (eqv? (expt 2 100) (expt 2 100))
+   (eqv? 2 2.0)
+   (let ([v (mcons 1 2)]) (eqv? v v))
+   (eqv? (mcons 1 2) (mcons 1 2))
+   (eqv? (integer->char 955) (integer->char 955))
+   (eqv? (make-string 3 #\z) (make-string 3 #\z))
+   (eqv? #t #t))}
+
+
+@defproc[(eq? [v1 any/c] [v2 any/c]) boolean?]{
+
+ Return @racket[#t] if @racket[v1] and @racket[v2] refer to the same
+ object, @racket[#f] otherwise. As a special case among @tech{numbers},
+ two @tech{fixnums} that are @racket[=] are also the same according
+ to @racket[eq?]. See also @secref["model-eq"].
+
+ @(examples
+   (eq? 'yes 'yes)
+   (eq? 'yes 'no)
+   (eq? (* 6 7) 42)
+   (eq? (expt 2 100) (expt 2 100))
+   (eq? 2 2.0)
+   (let ([v (mcons 1 2)]) (eq? v v))
+   (eq? (mcons 1 2) (mcons 1 2))
+   (eq? (integer->char 955) (integer->char 955))
+   (eq? (make-string 3 #\z) (make-string 3 #\z))
+   (eq? #t #t))}
+
+
+@defproc[
+ (equal?/recur [v1 any/c] [v2 any/c] [recur-proc (any/c any/c -> any/c)])
+ boolean?]{
+
+ Like @racket[equal?], but using @racket[recur-proc] for recursive
+ comparisons (which means that reference cycles are not handled
+ automatically). Non-@racket[#f] results from @racket[recur-proc] are
+ converted to @racket[#t] before being returned by
+ @racket[equal?/recur].
+
+ @(examples
+   (equal?/recur 1 1 (lambda (a b) #f))
+   (equal?/recur '(1) '(1) (lambda (a b) #f))
+   (equal?/recur '#(1 1 1) '#(1 1.2 3/4)
+                 (lambda (a b) (<= (abs (- a b)) 0.25))))}
+
+
+@section[#:tag "model-eq"]{Object Identity and Comparisons}
+
+
+The @racket[eq?] operator compares two @tech{values}, returning
+@racket[#t] when the values refer to the same @tech{object}. This form
+of equality is suitable for comparing objects that support imperative
+update (e.g., to determine that the effect of modifying an object
+through one reference is visible through another reference). Also, an
+@racket[eq?] test evaluates quickly, and @racket[eq?]-based hashing
+is more lightweight than @racket[equal?]-based hashing in hash tables.
+
```

Comment:
Is it perhaps worth adding some notes about contracts here? Something that explains that when an object passes through a contract boundary it may or may not come out @racket[eq?] on the other side?

Reply:
+ω

Reply:
Definitely. Trying to understand the relationships between `equal?`, `eq?`, `chaperone-of?`, and `impersonator-of?` and how that affects contracts is what motivated me to write this in the first place.

But I'd like to take it one step at a time and add that content in another pull request.

Reply:
Somethign more comprehensive would be great, but I'm advocating for a brief discussion that says that things that you think might be `eq?` because they are the "same object" is not so simple. That is, some might think that when you call a function with one object, you get "the same object" in the body, but when the function has a contract, you don't (always). More generally I think it is important to hint at the point that "same object" isn't a simple thing.

That said, if this pull request doesn't add any new content but just shuffles things, I'm happy to wait for a "new content" pull request to dig in then!

Reply:
Marking this resolved for now. I'll try to add a section about this in the next pull request.